notebooks/platform/xarray/lib/io/codecs.py

```diff
@@ -1,59 +1,60 @@
 
 
 class PackGeneticBits(PackBits):
-    """Custom Zarr plugin for encoding allele counts as 2 bit integers"""
-
-    codec_id = 'packgeneticbits'
+    """
+    Custom Zarr plugin for encoding allele counts as 2 bit integers.
 
-    def __init__(self):
-        super().__init__()
+    012 coding system must be used. 0 if the individual is homozygous for the more frequent allele,
+    1 if it is heterozygous, or 2 if it is homozygous for the less frequent allele. Use -1 for
+    missing data. Any other coding systems might lead to invalid encoding.
+    """
 
-    def encode(self, buf):
+    codec_id = "packgeneticbits"
+    __bit_mask = np.array([1, 2])
 
+    def encode(self, buf: np.ndarray) -> np.ndarray:
+        """Encodes a 2 dimensional array (variant x sample) of allele counts"""
         # normalise input
         arr = ensure_ndarray(buf)
+        if buf.ndim != 2:
+            raise ValueError("Input must be a 2 dimensional matrix")
 
         # broadcast to 3rd dimension having 2 elements, least significant
         # bit then second least, in that order; results in nxmx2 array
         # containing individual bits as bools
         # see: https://stackoverflow.com/questions/22227595/convert-integer-to-binary-array-with-suitable-padding
-        arr = arr[: ,: ,None] & (1 << np.arange(2)) > 0
-
-        return super().encode(arr)
+        arr = arr[:, :, None] & self.__bit_mask
+        bool_arr = arr > 0
+        # NOTE: -1 codes as [True, True] which equals to 3 in uint8 bit encoding
 
-    def decode(self, buf, out=None):
-
-        # normalise input
-        enc = ensure_ndarray(buf).view('u1')
+        return super().encode(bool_arr)
 
-        # flatten to simplify implementation
-        enc = enc.reshape(-1, order='A')
+    def decode(self, buf: np.ndarray, out: Optional[np.ndarray] = None) -> np.ndarray:
+        if buf.ndim != 1 or buf.dtype != np.uint8:
+            raise ValueError(
+                f"buf input must be a vector of uint8. buf shape: {buf.shape}, buf dtype: {buf.dtype}"
+            )
 
-        # find out how many bits were padded
-        n_bits_padded = int(enc[0])
-
-        # apply decoding
-        dec = np.unpackbits(enc[1:])
-
-        # remove padded bits
-        if n_bits_padded:
-            dec = dec[:-n_bits_padded]
-
-        # view as boolean array
-        dec = dec.view(bool)
+        dec = super().decode(buf, out=None)
 
         # given a flattened version of what was originally an nxmx2 array,
         # reshape to group adjacent bits in second dimension and
         # convert back to int based on each little-endian bit pair
-        dec = np.packbits(dec.reshape((-1, 2)), bitorder='little', axis=1)
+        dec = np.packbits(dec.reshape((-1, 2)), bitorder="little", axis=1)
         dec = dec.squeeze(axis=1)
 
+        # -1 which codes for missing data got encoded as 11000000 which codes for 3 in uint8,
+        # we revert that here
+        dec[dec == 3] = -1
```

Comment:
Thanks @eric-czech. FYI: this is the only change in the logic, the rest is cosmetic changes + tests.

Reply:
Makes sense, though I think this touches on something really important:

One of the early assumptions I made that turned out to be wrong was that it would be possible to use masked arrays with integer dtypes in xarray (https://github.com/pydata/xarray/issues/1194).  My plan from then on was to carry around the boolean mask array with the genotypes arrays and do math utilizing both until there's a fix for that.  I had intended for that to then mean that the encoding for the missing values in the genotypes array is more or less irrelevant and that it would be ok to skip whole-array traversals (which could be a significant slowdown given how few missing values there generally are) like this for the sake of preserving the -1 encoding.  My plan was then to use the boolean mask as the actual indicator of missingness in making all operations nan-aware.

The advantages of this are:
1. Making the missingness explicit and the sentinel values irrelevant (like np.ma) has a conceptual cohesiveness I like and is *probably* less error-prone
2. Allele counts are one kind of data type where a `-1` sentinel value for missing is somewhat standard and intuitive, but for some data types a choice for sentinel values is more awkward, e.g.:
    - allele indexes: -1 could mean last from the end
    - quantized allele dosages: demoting these to signed integers will be annoying
    - phasing information: This should be boolean but we would have to use signed ints instead
3. Operations can use the mask array for missingness and not need to be aware of the sentinel value strategy associated with each kind of data we want to support.  This is probably the biggest advantage. 

Disadvantages:
- The mask array occupies a lot of extra space, but:
  - Sparse arrays would be a good fit for them (they'll be at least ~80% False in most published datasets)
  - We can always do something like `x = xr.where(mask, sentinel_value, x)` if something needs to operate on a single array.  But again, then we need to make one of these sentinel choices for every kind of genetic data type outside of biallelic, diploid non-reference allele counts (which will be most common for a while, but not as we get more experienced).
- Some missingness patterns will be different within the same dataset.  For example, phasing may be missing for non-missing genotypes meaning that we would have to carry around multiple masks.

I should have started a larger discussion when you first pointed out the -1 issue, but here's what I think we need to decide:

- Do we always store data like this as uint knowing full well that the missing values will encode to some nonsense value, possibly within the domain of non-missing values?  It sounds dangerous but I'd argue that the danger introduced by needing a sentinel strategy for every single data type is similarly dangerous.

OR 

- Do we use -1 as the missing indicator in all integer data types, nan for missing in floats (which will confound with invalid calculations), only represent booleans as signed ints, and do away with masks entirely.

I don't like either of those options, and I apologize that a lot of this code isn't really committing hard one way or the other -- it's no doubt kind of confusing.  A third option is that we figure out how to support nullable integers in a single array.  That would solve a lot of these problems if it was possible.